### DISCUSSION TITLE
Дизайн функций работы с коллекциями

### DISCUSSION THREAD

**Discussion by gsvgit** [YaccConstructor/GraphBLAS-sharp#60] [Ideas]
Несколько мыслей по поводу того, как задизайнить внешний вид функций типа `map2`.
1. Добавить контекст, который будет доступен лямбде, которую мы туда передаём. Это должно помочь эмулировать захват внешних переменных лямбдами. Например при создании функции типа `assignByMask`. При этом, можно будет избежать лишних перекомпиляций кода ядер. Вероятно, это потребует поддержки на стороне Brahma.FSharp, чтобы типы везде красиво сходились.
2. Добавить таки буфер под результат явно, чтобы всегда снаружи явно управлять тем, с какими именно флагами должен быть выделен буфер под результаты и чтобы его можно было переиспользовать. Например, не аллоцировать реально новый фронт каждый раз, а гонять его между уже зааллоцированными буферами.
3.  Пример
```fsharp
let ctx = ClCell<int>()
let op = 
    <@
        fun ctx x y -> 
            match x,y with
            | None, None -> Some ctx.Value // вот тут, вероятно, надо будет в Brahma.FSharp что-то править
            | _ -> None 
   @>
let mutable front = ClArray()
let mutable newFront = ClArray()
map2 newFront op arr1 front
front <- newFrotn

```

**Comment by IgorErin** [comment]
Не совсем понимаю, как угадывать размер буфера при работе с разреженными структурами. Результат применения к ним `map2` может быть по размеру меньше исходного, в частном случае вообще пустым.
Или предполагается завести новую абстракцию, которая при себе будет хранить набор флагов, а при необходимости ее использования будет аллоцировать под себя соответствующее количество памяти? Это может решить проблему пустого результата.

**Reply by gsvgit** [reply]
> Не совсем понимаю, как угадывать размер буфера при работе с разреженными структурами. Результат применения к ним `map2` может быть по размеру меньше исходного, в частном случае вообще пустым. Или предполагается завести новую абстракцию, которая при себе будет хранить набор флагов, а при необходимости ее использования будет аллоцировать под себя соответствующее количество памяти? Это может решить проблему пустого результата.

Ну вот тут сложно, да. Для плотных размер знаем. Для разреженных нет. На крайний случай хотя бы флаги передавать, чтобы выделение контролировать.

С абстракцией тоже всё не так гладко. Не уверен, что получится красиво типы свести.

**Comment by IgorErin** [comment]
Относительно контекста тоже не понимаю. Пользователю придется писать операции где будет явное взаимодействие с абстрактным "контекстом"? 
Не совсем понимаю чем это отличается от нынешней реализации заполнения по маске для векторов. С учётом того что в текущей реализации работа с значением скрыта от пользователя. 
https://github.com/YaccConstructor/GraphBLAS-sharp/blob/d14ef08f35b9e392621e1533d315797f8ccad973/src/GraphBLAS-sharp.Backend/Quotes/Convert.fs#L15

**Reply by gsvgit** [reply]
> Относительно контекста тоже не понимаю. Пользователю придется писать операции где будет явное взаимодействие с абстрактным "контекстом"? Не совсем понимаю чем это отличается от нынешней реализации заполнения по маске для векторов. С учётом того что в текущей реализации работа с значением скрыта от пользователя.
> 
> https://github.com/YaccConstructor/GraphBLAS-sharp/blob/d14ef08f35b9e392621e1533d315797f8ccad973/src/GraphBLAS-sharp.Backend/Quotes/Convert.fs#L15

А круто придумано. 

Я, скорее, думаю о том, чтобы максимально всё обобщить и сдеать и маску через `map2` максимально честно и аккуратно.
